### DISCUSSION TITLE
Delete From Source Table in Destination Table with Subselect

### DISCUSSION THREAD

**Discussion by maticjosip** [sqlalchemy/sqlalchemy#9347] [Usage Questions] [answered]
As the source table and the destination table have the same name, i give an alias to the source table
```
# Renaming the source table name as issue occurs if no alias is given, as the table name is equal to the destination table name
src_model_alias = aliased(src_model, name='src_model_alias')
```

When i write the code like this the delete command works:
##### Option 1: Works
```
# DELETE records from destination table with same source table name and column names with 2 primary keys
subquery = (
	session
	.query(dest_model)
	.join(src_model_alias, (dest_model.PHARMACODE == src_model_alias.PHARMACODE) & (dest_model.LNO == src_model_alias.LNO)).subquery()
)

session\
	.query(dest_model)\
	.filter((dest_model.PHARMACODE == subquery.c.PHARMACODE) & (dest_model.LNO == subquery.c.LNO))\
	.delete(synchronize_session=False)
session.commit()
```

When i write a method for the same use case it does not:
##### Option 2: Does not work

```
def delete_matching_records(session, dest_model, src_model_alias, primary_key_names):
 # Dynamic DELETE method with Input variables, session, source table and destination table and primary key ID (list)

    # Construct the join condition
    subquery_join_condition = (" & ".join([f"{getattr(dest_model, col)} == {getattr(src_model_alias, col)}" for col in primary_key_names]))

    # Construct the subquery
    subquery = (
        session
        .query(dest_model)
        .join(src_model_alias, text(subquery_join_condition))
        .subquery()
    )

    del_join_condition = (" & ".join([f"({getattr(dest_model, col)} == {getattr(subquery.c, col)})" for col in primary_key_names]))
    # Output: 
    # (ArticleProprietaryQuantity.PHARMACODE == anon_1.PHARMACODE) & (ArticleProprietaryQuantity.LNO == anon_1.LNO)
    # ArticleProprietaryQuantity and anon_1 are the model_names 

    # Construct the delete query
    session\
        .query(dest_model)\
        .filter(del_join_condition)\
        .delete(synchronize_session=False)
	
	# Execute the delete command
    session.commit()
```
### Insights
1. The error message tells me to put the del_join_condition into text() to work
2. After adding the text() function this error arises, Incorrect syntax near '='. Wich is obvious as in the output the raw SQL Statement is shown `DELETE FROM [hospINDEX].[Article_Proprietary_Quantity] WHERE (ArticleProprietaryQuantity.PHARMACODE == anon_1.PHARMACODE) & (ArticleProprietaryQuantity.LNO == anon_1.LNO)]`
3. And when the double equal signs are replaced with single equal signs, the error occurs that the column names for the join could not be bound
4. As long as in the filter condition the filters are written out then it works but when I try to pass it via a variable it does not work

#### Test 1
`filter((model.PHARMACODE == subquery.c.PHARMACODE) & (model.LNO == subquery.c.LNO))`
#### Test 2
`filter(del_join_condition)`
#### Test 3
```
var = (model.PHARMACODE == subquery.c.PHARMACODE) & (model.LNO == subquery.c.LNO)
filter(var)
```

Any suggestions?

And thanks for your help!

**Comment by CaselIT** [comment, marked as answer]
Hi,

I'm not sure way are you trying to create a string of the SQL condition. If you are doing so you have to create correct sql, otherwise it's not gonna work. I would start by replacing `==` with `=` in this string you are creating. Also the escaping of the tables seems not correct.

If you just want to join on a dynamic primary key list, there is no need to create a sql string, a better option is to just use `sa,and_`
```py
del_join_condition = sa.and_(*(getattr(dest_model, col) == getattr(subquery.c, col) for col in primary_key_names))
```

**Reply by maticjosip** [reply]
Thanks for the quick reply! For your information, I am using SQLAlchemy 2.0.0

I am assuming sa is in your case the session. But when i try this i get the error message AttributeError: 'Session' object has no attribute 'and_'
Note: i have imported it with `from sqlalchemy import and_`

You are correct i want to have a method, which can take different src and dest tables with different number of primary keys, and delete all records with the same primary keys that exist in both tables in dest table -> The reason why is in the next method i would insert all records from th src to dest which results in sort of a delta load (Delete Insert)

**Reply by maticjosip** [reply]
My intention is to create a method that can replicate this code 
 ```
       subquery = (
            session
            .query(model)
            .join(stg_model_alias, (model.PHARMACODE == stg_model_alias.PHARMACODE) & (model.LNO == stg_model_alias.LNO)).subquery()
        )
        session\
            .query(model)\
            .filter((model.PHARMACODE == subquery.c.PHARMACODE) & (model.LNO == subquery.c.LNO))\
            .delete(synchronize_session=False)
        session.commit()
        delete_matching_records(session, model, stg_model_alias, id)
```

by passing src and dest table and a list of primary keys

**Reply by CaselIT** [reply]
> Note: i have imported it with from sqlalchemy import and_

yes, I've used the common convention `import sqlalchemy as sa`. Sorry.

Using that you can create a dynamic AND condition based on a list of keys, sure. No need to manually write sql

**Reply by maticjosip** [reply]
Awesome, now it works! Thanks for your help!
